add perfdash in CDK

mengqiy opened the pull request:
Added perfdash components (excluding configmap) in CDK.
Added a section in the README for perfdash.

hakuna-matatah commented on infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:
```diff
@@ -0,0 +1,118 @@
+    app.kubernetes.io/component: config
+    app.kubernetes.io/part-of: perfdash
+data:
+  jobs.yaml: |
```
- This data about jobs are more specific to internal things that we run. It doesn't make sense to keep this in upstream repo. Could we push this in our internal repo.
wdyt ?

mengqiy replied:
That makes sense.

hakuna-matatah commented on infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:
```diff
@@ -0,0 +1,118 @@
+            - /perfdash
+            - --www=true
+            - --address=0.0.0.0:8080
+            - --builds=20
```
IIUC does this mean we only display 20 builds on dashboard ?

mengqiy replied:
What's your suggestion? 50 or 100 or some other value?

hakuna-matatah commented on infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:
```diff
@@ -0,0 +1,118 @@
+            - --mode=s3
+            - --aws-region=us-west-2
+            - --configPath=/etc/config/jobs.yaml
+            - --logsBucket=kit-eks-scalability
```
Either we need to make sure we explicitly document this detail because other customers outside of us won't have access to this bucket and this deployment pods might fail or throw errors in the logs (OR) move this to internal repo ? Or patch this specific detail through flux  by taking in as param through cdk command line args ?

mengqiy replied:
I don't quite understand this comment.
Who is other customers? Other EKS teams or EKS customers?
IIUC this is bucket is only accessible to our team. Others can't access it no matter where we keep this yaml file.
Can you please elaborate your concern? We can discuss offline if you want.

mengqiy replied:
Updated according to our offline discussion.

hakuna-matatah commented on infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:
```diff
@@ -0,0 +1,118 @@
+            - --aws-region=us-west-2
+            - --configPath=/etc/config/jobs.yaml
+            - --logsBucket=kit-eks-scalability
+            - --logsPath=
```
any reason for having this here with empty value ?

mengqiy replied:
It appears upstream has another layer called `logs` in the hierarchy. We don't have that and it's not necessary for our use case. So we have it as an empty value.

hakuna-matatah replied:
cool, i see it here - https://github.com/kubernetes/perf-tests/blob/master/perfdash/perfdash.go#L52

mengqiy commented:
PTAL

mengqiy commented:
@hakuna-matatah Please note that I have created an internal CR for the ConfigMap. We need to merge and apply that before we merge this PR.

hakuna-matatah commented on infrastructure/README.md:
```diff
@@ -18,7 +18,52 @@ KIT Infrastructure creates a base K8s cluster with below  add-ons by default  bu
+To bring it up, there are a few places you need to ensure they are wired correctly.
+
+- The `logsBucket` logs bucket should be your bucket holding the logs. Ideally, it should be in the same AWS account as your KIT Infrastructure. Otherwise, you may need to configure the auth yourself.
+- A ConfigMap containing the jobs config for perf dash must be created so that it can be mounted as volumn in the Deployment. An example configmap can be:
```
Nit: volum -> volume

mengqiy replied:
Done

hakuna-matatah commented on infrastructure/README.md:
```diff
@@ -18,7 +18,52 @@ KIT Infrastructure creates a base K8s cluster with below  add-ons by default  bu
+
+To bring it up, there are a few places you need to ensure they are wired correctly.
+
+- The `logsBucket` logs bucket should be your bucket holding the logs. Ideally, it should be in the same AWS account as your KIT Infrastructure. Otherwise, you may need to configure the auth yourself.
```
Thanks for adding these details.

hakuna-matatah commented on infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml:
```diff
@@ -0,0 +1,81 @@
+            - /perfdash
+            - --www=true
+            - --address=0.0.0.0:8080
+            - --builds=100
```
Nice!

hakuna-matatah commented:
> @hakuna-matatah Please note that I have created an internal CR for the ConfigMap. We need to merge and apply that before we merge this PR.
> 
> 

Yup, thanks for moving that to internal.

mengqiy commented:
@hakuna-matatah  PTAL

hakuna-matatah reviewed (dismissed):
lgtm

mengqiy commented:
This PR should hold merging until the ConfigMap has been created out-of-the-band.

prateekgogia commented on infrastructure/README.md:
```diff
@@ -18,7 +18,52 @@ KIT Infrastructure creates a base K8s cluster with below  add-ons by default  bu
 - Kubernetes Iteration Toolkit (KIT) Operator (optional)
 
-Flux is setup, by deafult, to monitor the KIT git repo path `./infrastructure/k8s-config/clusters/kit-infrastructure`, which includes other add-ons that do not require AWS credentials such as tekton, prometheus, grafana, and the metrics-server. 
+Flux is setup, by deafult, to monitor the KIT git repo path `./infrastructure/k8s-config/clusters/kit-infrastructure`, which includes other add-ons that do not require AWS credentials such as tekton, prometheus, grafana, the metrics-server and [perf dash].
```
Curious why do we have perf dash text formatting like this - `[perf dash]`?

mengqiy replied:
It makes the raw markdown easier to read when the long URLs is not embedded in the text.
The link of `perf dash` is defined at the bottom of this file.

prateekgogia commented on infrastructure/README.md:
```diff
@@ -18,7 +18,52 @@ KIT Infrastructure creates a base K8s cluster with below  add-ons by default  bu
+
+#### Perf Dash
+
+Perf dash is optional. If you want to disable it, you can simply remove or comment out the content in `./infrastructure/k8s-config/clusters/kit-infrastructure/perfdash/perfdash.yaml`.
```
Doesn't provide info if its enabled by default? 
> If you want to disable it, you can simply remove or comment out the content in

I don't think this will work because it will be synced from Flux and we don't want everyone to be changing KIT repo for disabling perfdash. 

Also enabling by default would mean existing clusters in Karpenter will get this installed which might be a surprise for them. I would minimize such changes in this repo or at least get a consent from Karpenter team.
cc @bwagner5

mengqiy replied:
Thanks for bringing this up! It seems karpenter do use KIT for testing: https://github.com/aws/karpenter/tree/main/test/infrastructure#setup.

>I don't think this will work because it will be synced from Flux and we don't want everyone to be changing KIT repo for disabling perfdash.

I thought a user cloud have their own forks where they can make some changes. I can rephrase it to make it more clear. If you don't think that's a valid assumption, I can change it.

>Also enabling by default would mean existing clusters in Karpenter will get this installed which might be a surprise for them.

I guess it won't be a surprise for Karpenter team, since it doesn't break their cluster. But I guess this addon is not useful for them, since they don't CL2 tests at all. We may not want to install it default.
An alternative is to add it as an addon under `./infrastructure/lib/addons/` using typescript. A flag can be used to gate it and make it disabled by default.
@prateekgogia @hakuna-matatah WDYT?

hakuna-matatah replied:
Current state of things - We install certain addons today through typescript cdk thing if that addon requires aws resources to function otherwise we install them through flux. Customers have an option of disabling certain addons through cdk if they are being installed but not through flux.
We need to take a stand on whether we want to move all the addons to be installed through type-script and give customers an option of disabling unwanted addons (OR) have a separate directory or git repo to be pointed out for all optional addons so customers have flexibility to choose at the time of setup. For the latter one (i.e; flux route), all we have to do is basically send an array list here for repoPaths and repos, so customers can pass it at the time of setup  - https://github.com/awslabs/kubernetes-iteration-toolkit/blob/main/infrastructure/lib/kit-infrastructure.ts#L167-L177

mengqiy replied:
> have a separate directory or git repo to be pointed out for all optional addons so customers have flexibility to choose at the time of setup.

I guess this approach makes more sense.

prateekgogia commented on infrastructure/README.md:
```diff
@@ -69,4 +115,6 @@ cdk deploy KITInfrastructure --no-rollback \
\ No newline at end of file
+| AWSEBSCSIDriverAddon| AWSEBSCSIDriver CRD addon that gets installed on KIT Infrastructure by default             | true                                                    |   |   |
+
+[perf dash]: https://github.com/kubernetes/perf-tests/tree/master/perfdash
```
Why different formatting?

mengqiy replied:
Answered in https://github.com/awslabs/kubernetes-iteration-toolkit/pull/306#discussion_r1009719638

mengqiy commented:
Rebased on top of the latest HEAD. The last commit is the change since last review.
The changes is not fully tested due to https://github.com/awslabs/kubernetes-iteration-toolkit/issues/313. I will need some help to figure out what's going on with #313.

hakuna-matatah commented on infrastructure/lib/addons/fluxv2.ts:
```diff
@@ -10,10 +10,12 @@ export interface FluxV2Props extends StackProps {
   testRepoUrl?: string;
   testRepoBranch?: string;
   testRepoPath?: string;
+  testRepoAddonPaths?: string[];
```
why would you  need two arrays like one [here](https://github.com/awslabs/kubernetes-iteration-toolkit/pull/306/files#diff-bbf436cf80554d1c16a136a08cbc43d87d488d7ab9d51b6ca1aa66a74de555b9R13) and on this line ? Can't we just have one.

mengqiy replied:
From my current understanding, we support `Repo` and `testRepo` (this is not a great name, but we are using it now), user may want to install the addons from base `Repo` and (or) `testRepo`. We should provide both options IMO.

hakuna-matatah replied:
The idea there was to have `repo` for k8s addons that needs to be installed on Management cluster through Flux and `testRepo` is a place holder for tekton tasks tests etc that we wanted to separate out at that time from the Infrastructure code repository itself. 


Given with this PR, now we can take in array of repos, this logical separation of `repos` and `testRepos` may not be needed at this point. Because customers can bring in any number of arbitrary set of `testrepos` which contain tests to sync on cluster and/or also arbitrary set of k8s addons can be installed  `repos` using the same array list . Doesn't add any benefit now to have separate vars for this. Makes sense ?

hakuna-matatah commented on infrastructure/lib/addons/fluxv2.ts:
```diff
@@ -127,6 +129,30 @@ export class FluxV2 extends Construct {
 
+  if (props.repoAddonPaths) {
+    for (var repoAddonPath of props.repoAddonPaths) {
+      const addonKustomizationManifest = props.cluster.addManifest('AddonKustomizationSelf', {
```
could we wrap this in `function` as we are using it multiple times ?

mengqiy replied:
That's possible.
https://github.com/awslabs/kubernetes-iteration-toolkit/blob/da36db42fe000ace00c9396fb3abd0af226da3b0/infrastructure/lib/addons/fluxv2.ts#L110-L128
and https://github.com/awslabs/kubernetes-iteration-toolkit/blob/da36db42fe000ace00c9396fb3abd0af226da3b0/infrastructure/lib/addons/fluxv2.ts#L150-L168 are doing similar things for `Repo` and `testRepo`. We should clean them up as well eventually.

mengqiy commented:
@hakuna-matatah PTAL

hakuna-matatah reviewed (approved):
lgtm
